substrate/client/rpc-spec-v2/src/chain_head/tests.rs

```diff
@@ -4033,3 +4033,53 @@ async fn follow_event_with_unknown_parent() {
 	// When importing the block 2, chainHead detects a gap in our blocks and stops.
 	assert_matches!(get_next_event::<FollowEvent<String>>(&mut sub).await, FollowEvent::Stop);
 }
+
+#[tokio::test]
+async fn events_are_backpressured() {
+	let builder = TestClientBuilder::new();
+	let backend = builder.backend();
+	let client = Arc::new(builder.build());
+
+	let api = ChainHead::new(
+		client.clone(),
+		backend,
+		Arc::new(TokioTestExecutor::default()),
+		ChainHeadConfig {
+			global_max_pinned_blocks: MAX_PINNED_BLOCKS,
+			subscription_max_pinned_duration: Duration::from_secs(MAX_PINNED_SECS),
+			subscription_max_ongoing_operations: MAX_OPERATIONS,
+			max_lagging_distance: MAX_LAGGING_DISTANCE,
+			max_follow_subscriptions_per_connection: MAX_FOLLOW_SUBSCRIPTIONS_PER_CONNECTION,
+		},
+	)
+	.into_rpc();
+
+	let mut parent_hash = client.chain_info().genesis_hash;
+	let mut header = VecDeque::new();
+	let mut sub = api.subscribe("chainHead_v1_follow", [false], 511).await.unwrap();
+
+	// insert more events than the user can consume
+	for i in 0..=512 {
+		let block = BlockBuilderBuilder::new(&*client)
+			.on_parent_block(parent_hash)
+			.with_parent_block_number(i)
+			.build()
+			.unwrap()
+			.build()
+			.unwrap()
+			.block;
+		header.push_front(block.header().clone());
+
+		parent_hash = block.hash();
+		client.import(BlockOrigin::Own, block.clone()).await.unwrap();
+	}
+
+	let mut events = Vec::new();
+
+	while let Some(event) = sub.next::<FollowEvent<String>>().await {
+		events.push(event);
+	}
+
+	assert_eq!(events.len(), 64);
```

Comment:
dq: i'm not that familiar with testing libs but why subscription falls out at 63 instead of 511 here?
the error from overflowing buffer still happens tho 
cc: @niklasad1

Reply:
```bash
2024-10-15T15:07:45.783080Z TRACE jsonrpsee-server: [Methods::subscribe] Method: chainHead_v1_follow, params: Some(RawValue([false]))
2024-10-15T15:07:45.783324Z DEBUG rpc-spec-v2: [follow][id="8999421366389847"] Subscription accepted
2024-10-15T15:07:45.783524Z  INFO sc_rpc::utils: Starting subscription=chainHead_v1_followEvent
2024-10-15T15:07:45.783642Z TRACE jsonrpsee-server: [Methods::inner_call] Method: chainHead_v1_follow, response: {"jsonrpc":"2.0","id":0,"result":8999421366389847}
2024-10-15T15:07:47.380636Z DEBUG jsonrpsee-server: [Subscription::next]: rx {"jsonrpc":"2.0","method":"chainHead_v1_followEvent","params":{"subscription":8999421366389847,"result":{"event":"initialized","finalizedBlockHashes":["0xb4b48648e4869cffd7333c79458c3db011b9d6556af2a53eb5a186763837a73c"]}}}
2024-10-15T15:07:47.382317Z DEBUG rpc-spec-v2: [follow][id="8999421366389847"] Failed to handle stream notification ExceededLimits
2024-10-15T15:07:47.382353Z DEBUG rpc-spec-v2: [follow][id="8999421366389847"] Failed to handle stream notification SubscriptionAbsent
2024-10-15T15:07:47.383164Z DEBUG rpc-spec-v2: [follow][id="8999421366389847"] Subscription removed
```

It looks like you are hitting `max block pinning limits` when you import so many blocks.
Thus, it may be better to just add `subscription buffer cap` to the ChainHeadConfig to make it easy to change for tests etc.

Reply:
added a s`ubscription buffer cap` to the chainhead config, now we use it inside generate_events